shell: add an option for minimal default configuration

jackrosenthal opened the pull request:
The default shell configuration has heavy flash and memory requirements,
requiring project maintainers to set many configuration options to "n"
to keep flash and memory requirements within reason.

This adds a new configuration option, CONFIG_SHELL_MINIMAL, which
will disable flash and memory heavy options by default, and allow
project maintainers to select/imply only the options they want.

On a quick test from an ARM board I'm working on, enabling this option
cut flash space requirements by ~8 KB, and memory requirements by ~1 KB.

Signed-off-by: Jack Rosenthal <jrosenth@chromium.org>

jackrosenthal committed:
shell: add an option for minimal default configuration

The default shell configuration has heavy flash and memory requirements,
requiring project maintainers to set many configuration options to "n"
to keep flash and memory requirements within reason.

This adds a new configuration option, CONFIG_SHELL_MINIMAL, which will
disable flash and memory heavy options by default, and allow project
maintainers to select/imply only the options they want.

On a quick test from an ARM board I'm working on, enabling this option
cut flash space requirements by ~8 KB, and memory requirements by ~1 KB.

Signed-off-by: Jack Rosenthal <jrosenth@chromium.org>

sjg20 commented on subsys/shell/Kconfig:
```diff
@@ -61,12 +69,14 @@ config SHELL_DEFAULT_TERMINAL_HEIGHT
 config SHELL_TAB
 	bool "Enable the Tab button support in shell"
+	default n if SHELL_LIGHTWEIGHT
 	default y
```
default y if !SHELL_LIGHTWEIGHT

similarly below

It seems odd to have 'default n' when that is the default anyway, until otherwise chosen

jackrosenthal replied:
Done

sjg20 commented on subsys/shell/Kconfig:
```diff
@@ -17,6 +17,14 @@ module-str = Shell
 source "subsys/logging/Kconfig.template.log_config"
 source "subsys/shell/Kconfig.backends"
 
+config SHELL_LIGHTWEIGHT
```
would SHELL_MINIMAL be a better name?

jackrosenthal replied:
why MINIMAL over LIGHTWEIGHT?

sjg20 replied:
Well it seems to be the pattern in Zephyr, e.g. MINIMAL_LIBC

git grep MINIMAL

jfischer-no replied:
+1 for MINIMAL

e.g. LOG_MODE_MINIMAL

jackrosenthal replied:
Why MINIMAL vs. LIGHTWEIGHT?

keith-zephyr replied:
I see both uses in the Zephyr code, but "minimal" seems to be more widely adopted:  e.g. CONFIG_MINIMAL_LIBC, CONFIG_LOG_MINIMAL, LOG_MODE_MINIMAL.  I vote for SHELL_MINIMAL

jackrosenthal replied:
Done

sjg20 commented on subsys/shell/Kconfig.backends:
```diff
@@ -135,6 +136,7 @@ if SHELL_BACKEND_RTT
 
 config SHELL_PROMPT_RTT
 	string "Displayed prompt name"
+	default "> " if SHELL_LIGHTWEIGHT
```
Is it worth changing this and the ones below?

They do actually provide useful info
Plus they are probably not used in most boards - e.g. dummy is really just for tests I think

jackrosenthal replied:
Snuck this in here trying to get the prompt to look like something FAFT likes ... but I should make a different configuration option for that ... so dropping it

jakub-uC commented:
@jackrosenthal 
Would you mind update documentation about this configuration option?
On top of that, it might be good to add `CONFIG_SHELL_MINIMAL`config file in tests here:
https://github.com/zephyrproject-rtos/zephyr/tree/master/tests/subsys/shell/shell
and update testcase.yaml accordingly.

jackrosenthal commented:
@jakub-uC  Sure, I updated the documentation and the testcases accordingly.

For the test cases, they generally had the pattern of setting all of the config options to `n` except for the options desired.  I changed them to use `CONFIG_SHELL_MINIMAL=y` and enable the options they required instead.  Let me know if this is acceptable.
